virtool/analyses/data.py

```diff
@@ -1,23 +1,391 @@
+
+        if right == "read":
+            return read
+
+        if right == "write":
+            return write
+
+    async def delete(self, analysis_id: str, jobs_api_flag: bool):
+        """
+        Delete a single analysis by its ID.
+
+        :param analysis_id: the analysis ID
+        :param jobs_api_flag: checks if the jobs_api is handling the request
+        """
+        document = await self._db.analyses.find_one(
+            {"_id": analysis_id}, ["job", "ready", "sample"]
+        )
+
+        if not document:
+            raise ResourceNotFoundError()
+
+        sample_id = document["sample"]["id"]
+
+        if jobs_api_flag:
+            if document["ready"]:
+                raise ResourceConflictError()
+        else:
+            if not document["ready"]:
+                raise ResourceConflictError()
+
+        await self._db.analyses.delete_one({"_id": analysis_id})
+
+        path = self._config.data_path / "samples" / sample_id / "analysis" / analysis_id
+
+        try:
+            await run_in_thread(rm, path, True)
+        except FileNotFoundError:
+            pass
+
+        await recalculate_workflow_tags(self._db, sample_id)
+
+    async def upload(
+        self, reader, analysis_id: str, analysis_format: str, name: str
+    ) -> Optional[AnalysisFile]:
+        """
+        Uploads a new analysis result file.
+
+        :param reader: the file reader
+        :param analysis_id: the analysis ID
+        :param analysis_format: the format of the analysis
+        :param name: the name of the analysis file
+        :return: the new analysis file
+        """
+
+        document = await self._db.analyses.find_one(analysis_id)
+
+        if document is None:
+            raise ResourceNotFoundError()
+
+        analysis_file = await create_analysis_file(
+            self._pg, analysis_id, analysis_format, name
+        )
+
+        upload_id = analysis_file["id"]
+
+        analysis_file_path = (
+            self._config.data_path / "analyses" / analysis_file["name_on_disk"]
+        )
+
+        try:
+            size = await naive_writer(reader, analysis_file_path)
+        except CancelledError:
+            logger.debug(f"Analysis file upload aborted: {upload_id}")
+            await delete_row(self._pg, upload_id, AnalysisFile)
+
+            return None
+
+        analysis_file = await virtool.uploads.db.finalize(
+            self._pg, size, upload_id, AnalysisFile
+        )
+
+        return AnalysisFile(**analysis_file)
+
+    async def get_file_name(self, upload_id: int) -> str:
+        """
+        Get a file generated during the analysis.
+
+        :param upload_id: the upload ID
+        :return: the name on disk of the analysis file
+        """
+
+        analysis_file = await get_row_by_id(self._pg, AnalysisFile, upload_id)
+
+        if analysis_file:
+            return analysis_file.name_on_disk
+
+        raise ResourceNotFoundError()
+
+    async def download_analysis(
```

Comment:
This method is on the analysis data layer so it is not necessary to be so detailed with this name.
```suggestion
    async def download(
```